### PR TITLE
Add immutable strings array library

### PR DESCRIPTION
**Pull request by ryley-o**
## Description of the change

Adds an immutable strings library to benefit future systems.

fixes https://linear.app/art-blocks/issue/PRO-1104/immutable-strings-array-solidity-library

This adds a library that efficiently uses contract bytecode to efficiently store immutable arrays of strings. This massively reduces gas costs when storing and loading string arrays.

Assembly is thoroughly commented and simplified for readability.

## Tests

Tests are added for coverage of the new functionality, as well as edge cases.

### CODE REVIEW

**Review by lindsgil — APPROVED** [review]
🔥 🔥 🔥

File: packages/contracts/contracts/libs/v0.8.x/ImmutableStringArray.sol [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+        uint256 packedDataLength = 8 +
+            (arrayLength * 8) +
+            totalStringBytesLength;
+        bytes memory packedData = new bytes(packedDataLength + 0x20);
+        uint256 ptr;
+
+        // store the length of the strings array in the first 8 bytes of the packedData
+        assembly ("memory-safe") {
+            ptr := add(packedData, 0x20) // pointer to first byte of packedData
+            let existing := mload(ptr) // load the full 32-byte word
```
Comment (yoshiwarab) [inline packages/contracts/contracts/libs/v0.8.x/ImmutableStringArray.sol:63]: Do we need this?

**Review by yoshiwarab — APPROVED** [review]

File: packages/contracts/contracts/libs/v0.8.x/ImmutableStringArray.sol (hunk 1 above)
Comment (ryley-o) [reply packages/contracts/contracts/libs/v0.8.x/ImmutableStringArray.sol]: good catch - removed in [4cc824e](https://github.com/ArtBlocks/artblocks-contracts/pull/1767/commits/4cc824e9a82d0e850c23e767e52f65ed6856196f)
